locator/token_metadata.cc

```diff
@@ -769,122 +719,113 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
         keyspace_name, _bootstrap_tokens, _leaving_endpoints, _replacing_endpoints);
     if (_bootstrap_tokens.empty() && _leaving_endpoints.empty() && _replacing_endpoints.empty()) {
         tlogger.debug("No bootstrapping, leaving nodes, replacing nodes -> empty pending ranges for {}", keyspace_name);
-        set_pending_ranges(keyspace_name, std::unordered_multimap<range<token>, inet_address>(), can_yield::no);
+        _pending_ranges_interval_map.erase(keyspace_name);
         return make_ready_future<>();
     }
 
     return async([this, &unpimplified_this, &strategy, keyspace_name, &get_dc_rack] () mutable {
-        std::unordered_multimap<range<token>, inet_address> new_pending_ranges;
-        calculate_pending_ranges_for_replacing(unpimplified_this, strategy, new_pending_ranges);
-        // Copy of metadata reflecting the situation after all leave operations are finished.
-        auto all_left_metadata = make_token_metadata_ptr(clone_after_all_left().get0());
-        calculate_pending_ranges_for_leaving(unpimplified_this, strategy, new_pending_ranges, all_left_metadata);
-        // At this stage newPendingRanges has been updated according to leave operations. We can
-        // now continue the calculation by checking bootstrapping nodes.
-        calculate_pending_ranges_for_bootstrap(strategy, new_pending_ranges, all_left_metadata, get_dc_rack);
-        all_left_metadata->clear_gently().get();
-
-        // At this stage newPendingRanges has been updated according to leaving and bootstrapping nodes.
-        set_pending_ranges(keyspace_name, std::move(new_pending_ranges), can_yield::yes);
+        // new_token_metadata incorporates all the changes from leaving, bootstrapping and replacing
+        const auto new_token_metadata = token_metadata([&]() -> std::unique_ptr<token_metadata_impl> {
+            auto result = clone_only_token_map(false).get0();
+
+            // construct new_normal_tokens based on _bootstrap_tokens and _replacing_endpoints
+            std::unordered_map<inet_address, std::unordered_set<token>> new_normal_tokens;
+            if (!_replacing_endpoints.empty()) {
+                for (const auto& [token, inet_address]: _token_to_endpoint_map) {
+                    const auto it = _replacing_endpoints.find(inet_address);
+                    if (it == _replacing_endpoints.end()) {
+                        continue;
+                    }
+                    new_normal_tokens[it->second].insert(token);
+                }
+                for (const auto& [replace_from, replace_to]: _replacing_endpoints) {
+                    result->remove_endpoint(replace_from);
+                }
+            }
+            {
+                std::unordered_map<inet_address, std::unordered_set<token>> bootstrap_tokens;
+                for (const auto& [token, inet_address]: _bootstrap_tokens) {
+                    bootstrap_tokens[inet_address].insert(token);
```

Comment:
> replacing e1 with e2 and bootstrap e2 at the same time

This is the only possible way to replace. You cannot perform a replace without bootstrapping a node.

But what's the "collision" in this scenario? The tokens that the replacing node will own are exactly the tokens that the replaced node has owned. There are no additional tokens introduced. `_bootstrap_tokens` should be empty in that case and if not, they should be exactly the same as what you find in the previous loop.

Anyway what I meant is that there must not be two different nodes e2 and e3 such that they are introducing the same tokens (either by replacing or by bootstrapping).

Reply:
> But what's the "collision" in this scenario?

```
replace(e1 -> e2)
bootstrap(e2, {e2_tokens})
```

The conflict is in what tokens should be assigned to e2, those that was initially assigned to e1, or e2_tokens.

Maybe my confusion is that I don't fully understand why we need replace in the first place and what are its usage scenarios.

Reply:
>  The conflict is in what tokens should be assigned to e2, those that was initially assigned to e1, or e2_tokens.

`e2` does not introduce any tokens on its own during replace, it takes `e1`'s tokens. There is no way to both introduce new tokens and steal someone else's tokens currently. If there was we would have to take the sum of both when calculating `new_normal_tokens`.

Reply:
> Maybe my confusion is that I don't fully understand why we need replace in the first place and what are its usage scenarios.

It's faster to perform a replace operation than to `removenode` and then follow-up with bootstrap.

Reply:
1. The issue seems not particularly important, since these conflicts can't occur in practice, if I understand correctly. We need some consistent behaviour with min scope of change.
2. From the `token_metadata` point of view we can treat `replace(e1, e2)` just as a shortcut for 
`bootstrap(e2, e1_tokens); leave(e1)`;
3. Currently we can call `bootstrap(e, <some tokens>)` many times, the effect is that the tokens are merged.

So it seems ok to just merge tokens in our case too. This means I can drop `bootstrap_tokens` and update `new_normal_tokens` directly, as you proposed.